x/ccv/provider/keeper/distribution.go

```diff
@@ -57,32 +66,273 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
-				ctx,
-				types.ConsumerRewardsPool,
-				k.feeCollectorName,
-				sdk.NewCoins(balance),
+		// allocate remaining rewards to the community pool
+		remainingRewards, remainingChanges := remaining.TruncateDecimal()
+		err = k.distributionKeeper.FundCommunityPool(context.Context(ctx), remainingRewards, k.accountKeeper.GetModuleAccount(ctx, types.ConsumerRewardsPool).GetAddress())
+		if err != nil {
+			k.Logger(ctx).Error(
+				"fail to allocate rewards from consumer chain %s to community pool: %s",
+				consumerChainID,
+				err,
 			)
-			if err != nil {
-				k.Logger(ctx).Error("cannot sent consumer rewards to fee collector:", "reward", balance.String())
-			}
+			continue
+		}
+
+		// set consumer allocations to the remaining rewards decimals
+		alloc.Rewards = validatorsRewardsChange.Add(remainingChanges...)
+		k.SetConsumerRewardsAllocation(ctx, consumerChainID, alloc)
+	}
+}
+
+// IsEligibleForConsumerRewards returns `true` if the validator with `consumerValidatorHeight` has been a consumer
+// validator for a long period of time and hence is eligible to receive rewards, and false otherwise
+func (k Keeper) IsEligibleForConsumerRewards(ctx sdk.Context, consumerValidatorHeight int64) bool {
+	numberOfBlocksToStartReceivingRewards := k.GetNumberOfEpochsToStartReceivingRewards(ctx) * k.GetBlocksPerEpoch(ctx)
+
+	// a validator is eligible for rewards if it has been a consumer validator for `NumberOfEpochsToStartReceivingRewards` epochs
+	return (ctx.BlockHeight() - consumerValidatorHeight) >= numberOfBlocksToStartReceivingRewards
+}
+
+// AllocateTokensToConsumerValidators allocates tokens
+// to the given consumer chain's validator set
+func (k Keeper) AllocateTokensToConsumerValidators(
+	ctx sdk.Context,
+	chainID string,
+	tokens sdk.DecCoins,
+) (allocated sdk.DecCoins) {
+	// return early if the tokens are empty
+	if tokens.Empty() {
+		return allocated
+	}
+
+	// get the total voting power of the consumer valset
+	totalPower := math.LegacyNewDec(k.ComputeConsumerTotalVotingPower(ctx, chainID))
+	if totalPower.IsZero() {
+		return allocated
+	}
+
+	// Allocate tokens by iterating over the consumer validators
+	for _, consumerVal := range k.GetConsumerValSet(ctx, chainID) {
+		// if a validator is not eligible, this means that the other eligible validators would get more rewards
+		if !k.IsEligibleForConsumerRewards(ctx, consumerVal.JoinHeight) {
+			continue
 		}
+
+		consAddr := sdk.ConsAddress(consumerVal.ProviderConsAddr)
+
+		// get the validator tokens fraction using its voting power
+		powerFraction := math.LegacyNewDec(consumerVal.Power).QuoTruncate(totalPower)
+		tokensFraction := tokens.MulDecTruncate(powerFraction)
+
+		// get the validator type struct for the consensus address
+		val, err := k.stakingKeeper.GetValidatorByConsAddr(ctx, consAddr)
+		if err != nil {
+			k.Logger(ctx).Error("cannot find validator by consensus address :%s while allocating rewards from consumer chain: %s",
```

Comment:
Here and below: Does string formatting (e.g., `%s`) work together with `k.Logger(ctx)`?

Reply:
It doesn't, good catch.